osu.Game.Rulesets.Osu/Skinning/Smoke.cs

```diff
@@ -0,0 +1,445 @@
+        }
+
+        protected abstract override DrawNode CreateDrawNode();
+
+        protected override void Update()
+        {
+            base.Update();
+
+            Position = TopLeft;
+        }
+
+        protected override void Dispose(bool isDisposing)
+        {
+            base.Dispose(isDisposing);
+
+            if (smokeContainer != null)
+            {
+                smokeContainer.SmokeMoved -= onSmokeMoved;
+                smokeContainer.SmokeEnded -= onSmokeEnded;
+            }
+        }
+
+        protected struct SmokePoint
+        {
+            public Vector2 Position;
+            public double Time;
+            public Vector2 Direction;
+
+            public struct UpperBoundComparer : IComparer<SmokePoint>
+            {
+                public int Compare(SmokePoint x, SmokePoint target)
+                {
+                    // By returning -1 when the target value is equal to x, guarantees that the
+                    // element at BinarySearch's returned index will always be the first element
+                    // larger. Since 0 is never returned, the target is never "found", so the return
+                    // value will be the index's complement.
+
+                    return x.Time > target.Time ? 1 : -1;
+                }
+            }
+        }
+
+        protected abstract class SmokeDrawNode : TexturedShaderDrawNode
+        {
+            protected new Smoke Source => (Smoke)base.Source;
+
+            protected IVertexBatch<TexturedVertex2D>? QuadBatch;
+
+            protected readonly List<SmokePoint> Points = new List<SmokePoint>();
+
+            protected float Radius;
+            protected Vector2 DrawSize;
+            protected Vector2 PositionOffset;
+            protected Texture? Texture;
+
+            protected double SmokeStartTime;
+            protected double SmokeEndTime;
+            protected double CurrentTime;
+
+            protected RectangleF TextureRect;
+
+            private IFrameBasedClock? clock;
+
+            protected SmokeDrawNode(ITexturedShaderDrawable source)
+                : base(source)
+            {
+            }
+
+            public override void ApplyState()
+            {
+                base.ApplyState();
+
+                Points.Clear();
+                Points.AddRange(Source.SmokePoints);
+
+                Radius = Source.Radius;
+                DrawSize = Source.DrawSize;
+                PositionOffset = Source.TopLeft;
+                Texture = Source.Texture;
+                clock = Source.Clock;
+
+                SmokeStartTime = Source.SmokeStartTime;
+                SmokeEndTime = Source.SmokeEndTime;
+            }
+
+            public sealed override void Draw(IRenderer renderer)
+            {
+                base.Draw(renderer);
+
+                if (Points.Count == 0)
+                    return;
+
+                QuadBatch ??= renderer.CreateQuadBatch<TexturedVertex2D>(7200, 10);
```

Comment:
Probably need to cap smoke to the size of this VBO or possibly even less, since you can lag the game by drawing with no limit.

Technically this is a prime case for instanced rendering which can improve things in the future, but things are in a bit of a state of flux at the moment.